src/Compilers/CSharp/Test/Syntax/Parsing/PatternParsingTests_ListPatterns.cs

```diff
@@ -0,0 +1,722 @@
+            {
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "c");
+                }
+                N(SyntaxKind.IsKeyword);
+                N(SyntaxKind.RecursivePattern);
+                {
+                    N(SyntaxKind.LengthPatternClause);
+                    {
+                        N(SyntaxKind.OpenBracketToken);
+                        N(SyntaxKind.RelationalPattern);
+                        {
+                            N(SyntaxKind.GreaterThanToken);
+                            N(SyntaxKind.NumericLiteralExpression);
+                            {
+                                N(SyntaxKind.NumericLiteralToken, "0");
+                            }
+                        }
+                        N(SyntaxKind.CloseBracketToken);
+                    }
+                }
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void LengthPattern_03()
+        {
+            UsingExpression(@"c is [_]{P:P}");
+
+            N(SyntaxKind.IsPatternExpression);
+            {
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "c");
+                }
+                N(SyntaxKind.IsKeyword);
+                N(SyntaxKind.RecursivePattern);
+                {
+                    N(SyntaxKind.LengthPatternClause);
+                    {
+                        N(SyntaxKind.OpenBracketToken);
+                        N(SyntaxKind.DiscardPattern);
+                        {
+                            N(SyntaxKind.UnderscoreToken);
+                        }
+                        N(SyntaxKind.CloseBracketToken);
+                    }
+                    N(SyntaxKind.PropertyPatternClause);
+                    {
+                        N(SyntaxKind.OpenBraceToken);
+                        N(SyntaxKind.Subpattern);
+                        {
+                            N(SyntaxKind.NameColon);
+                            {
+                                N(SyntaxKind.IdentifierName);
+                                {
+                                    N(SyntaxKind.IdentifierToken, "P");
+                                }
+                                N(SyntaxKind.ColonToken);
+                            }
+                            N(SyntaxKind.ConstantPattern);
+                            {
+                                N(SyntaxKind.IdentifierName);
+                                {
+                                    N(SyntaxKind.IdentifierToken, "P");
+                                }
+                            }
+                        }
+                        N(SyntaxKind.CloseBraceToken);
+                    }
+                }
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void LengthPattern_04()
+        {
+            UsingExpression(@"c is ()[0]");
```

Comment:
nit: I still struggle with this combination... I have no intuition on whether the positional-pattern should come before or after the length-pattern :-/

Reply:
We could possibly move the length pattern before positionals, you'd have `T[N]`, `[N] { ... }`, `[N] ( ... )`, etc.